File: Server/ClientHandler.cs
using System.Net.Sockets;

namespace AircraftFuelServer;

/// <summary>
/// Handles communication with a single connected TCP client (aircraft).
/// Reads telemetry lines, parses packets, and feeds data to FleetDataManager.
/// </summary>
public static class ClientHandler
{
    /// <summary>
    /// Main handler loop for a connected client.
    /// Reads pipe-delimited telemetry packets line by line until disconnect or EOF.
    /// On completion, finalizes the flight record in the fleet manager.
    /// </summary>
    public static void HandleClient(TcpClient client, FleetDataManager manager)
    {
        string? airplaneId = null;
        NetworkStream? stream = null;
        StreamReader? reader = null;

        try
        {
            stream = client.GetStream();
            reader = new StreamReader(stream);

            Logger.Log($"Client connected from {client.Client.RemoteEndPoint}");

            string? line;
            while ((line = reader.ReadLine()) != null)
            {
                if (string.IsNullOrWhiteSpace(line))
                    continue;

                try
                {
                    var (id, dataPoint) = ParsePacket(line);
                    airplaneId = id;
                    manager.UpdateFuelReading(id, dataPoint.Timestamp, dataPoint.FuelRemaining);
                }
                catch (FormatException ex)
                {
                    Logger.Log($"[ClientHandler] Malformed packet: '{line}' — {ex.Message}");
                }
            }

            Logger.Log($"Client {airplaneId ?? "unknown"} disconnected (EOF).");
        }
        catch (IOException ex)
        {
            Logger.Log($"[ClientHandler] IO error for {airplaneId ?? "unknown"}: {ex.Message}");
        }
        catch (SocketException ex)
        {
            Logger.Log($"[ClientHandler] Socket error for {airplaneId ?? "unknown"}: {ex.Message}");
        }
        finally
        {
            if (airplaneId != null)
                manager.CompleteFlight(airplaneId);

            reader?.Dispose();
            stream?.Dispose();
            client?.Dispose();

            Logger.Log($"Connection for {airplaneId ?? "unknown"} fully closed.");
        }
    }

    /// <summary>
    /// Parses a pipe-delimited telemetry packet of the form:
    ///   AIRPLANE_ID|TIMESTAMP|FUEL_REMAINING
    /// Returns the airplane ID and a TelemetryDataPoint.
    /// Throws FormatException if the packet is malformed.
    /// </summary>
    public static (string AirplaneId, TelemetryDataPoint DataPoint) ParsePacket(string line)
    {
        string[] parts = line.Split('|');
        if (parts.Length != 3)
            throw new FormatException($"Expected 3 pipe-delimited fields, got {parts.Length}.");

        string id = parts[0].Trim();
        if (string.IsNullOrEmpty(id))
            throw new FormatException("Airplane ID is empty.");

        if (!DateTime.TryParse(parts[1].Trim(), out DateTime timestamp))
            throw new FormatException($"Cannot parse timestamp: '{parts[1]}'.");

        if (!double.TryParse(parts[2].Trim(), System.Globalization.NumberStyles.Any,
                System.Globalization.CultureInfo.InvariantCulture, out double fuelRemaining))
            throw new FormatException($"Cannot parse fuel value: '{parts[2]}'.");

        return (id, new TelemetryDataPoint { Timestamp = timestamp, FuelRemaining = fuelRemaining });
    }
}

File: Server/ClientHandler.cpp
#include <winsock2.h>
#include "ClientHandler.h"
#include "Logger.h"
#include <string>
#include <sstream>

bool ClientHandler::parsePacket(const std::string& line, std::string& airplaneId,
                                 std::string& timestamp, double& fuelRemaining) {
    // Expected format: "AIRPLANE_ID|TIMESTAMP|FUEL_REMAINING"
    if (line.empty()) return false;

    // Find first '|'
    std::size_t pos1 = line.find('|');
    if (pos1 == std::string::npos) return false;

    // Find second '|'
    std::size_t pos2 = line.find('|', pos1 + 1);
    if (pos2 == std::string::npos) return false;

    airplaneId = line.substr(0, pos1);
    timestamp  = line.substr(pos1 + 1, pos2 - pos1 - 1);

    std::string fuelStr = line.substr(pos2 + 1);
    // Trim any trailing whitespace/CR
    while (!fuelStr.empty() && (fuelStr.back() == '\r' || fuelStr.back() == '\n' || fuelStr.back() == ' ')) {
        fuelStr.pop_back();
    }

    if (airplaneId.empty() || timestamp.empty() || fuelStr.empty()) return false;

    try {
        fuelRemaining = std::stod(fuelStr);
    } catch (...) {
        return false;
    }

    return true;
}

void ClientHandler::handleClient(SOCKET clientSocket, FleetDataManager* manager) {
    Logger::log("Client handler started for new connection.");

    std::string lastAirplaneId;
    std::string lineBuffer;

    while (true) {
        char ch = 0;
        int result = recv(clientSocket, &ch, 1, 0);

        if (result == 0) {
            // Graceful disconnect
            Logger::log("Client disconnected gracefully.");
            break;
        } else if (result == SOCKET_ERROR) {
            Logger::log("Socket error while receiving data.");
            break;
        }

        if (ch == '\n') {
            // Process the line
            // Strip trailing '\r' if present (CRLF line endings)
            if (!lineBuffer.empty() && lineBuffer.back() == '\r') {
                lineBuffer.pop_back();
            }

            if (!lineBuffer.empty()) {
                std::string airplaneId, timestamp;
                double fuelRemaining = 0.0;

                if (parsePacket(lineBuffer, airplaneId, timestamp, fuelRemaining)) {
                    lastAirplaneId = airplaneId;
                    manager->updateFuelReading(airplaneId, timestamp, fuelRemaining);
                } else {
                    Logger::log("Malformed packet received: " + lineBuffer);
                }
            }

            lineBuffer.clear();
        } else {
            lineBuffer += ch;
        }
    }

    // Finalize the flight for this client
    if (!lastAirplaneId.empty()) {
        Logger::log("Completing flight for aircraft: " + lastAirplaneId);
        manager->completeFlight(lastAirplaneId);
    }

    closesocket(clientSocket);
    Logger::log("Client socket closed.");
}
